smart_open/gcs.py

```diff
@@ -0,0 +1,523 @@
+            client=None,  # type: google.cloud.storage.Client
+    ):
+        if client is None:
+            client = google.cloud.storage.Client()
+        self._client = client
+        self._credentials = self._client._credentials
+        self._bucket = self._client.bucket(bucket)  # type: google.cloud.storage.Bucket
+        self._blob = self._bucket.blob(blob)  # type: google.cloud.storage.Blob
+        assert min_part_size % _MIN_MIN_PART_SIZE == 0, 'min part size must be a multiple of 256KB'
+        self._min_part_size = min_part_size
+
+        self._total_size = 0
+        self._total_parts = 0
+        self._buf = io.BytesIO()
+
+        self._session = google_requests.AuthorizedSession(self._credentials)
+
+        #
+        # https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#start-resumable
+        #
+        self._resumeable_upload_url = self._blob.create_resumable_upload_session()
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    def flush(self):
+        pass
+
+    #
+    # Override some methods from io.IOBase.
+    #
+    def close(self):
+        logger.debug("closing")
+        if self._total_size == 0:  # empty files
+            self._upload_empty_part()
+        if self._buf.tell():
+            self._upload_next_part()
+        logger.debug("successfully closed")
+
+    def writable(self):
+        """Return True if the stream supports writing."""
+        return True
+
+    def tell(self):
+        """Return the current stream position."""
+        return self._total_size
+
+    #
+    # io.BufferedIOBase methods.
+    #
+    def detach(self):
+        raise io.UnsupportedOperation("detach() not supported")
+
+    def write(self, b):
+        """Write the given bytes (binary string) to the GCS file.
+
+        There's buffering happening under the covers, so this may not actually
+        do any HTTP transfer right away."""
+
+        if not isinstance(b, _BINARY_TYPES):
+            raise TypeError(
+                "input must be one of %r, got: %r" % (_BINARY_TYPES, type(b)))
+
+        self._buf.write(b)
+        self._total_size += len(b)
+
+        if self._buf.tell() >= self._min_part_size:
+            self._upload_next_part()
+
+        return len(b)
+
+    def terminate(self):
+        #
+        # https://cloud.google.com/storage/docs/xml-api/resumable-upload#example_cancelling_an_upload
+        #
+        self._session.delete(self._resumeable_upload_url)
+
+    #
+    # Internal methods.
+    #
+    def _upload_next_part(self):
+        part_num = self._total_parts + 1
+        logger.info("uploading part #%i, %i bytes (total %.3fGB)",
+                    part_num, self._buf.tell(), self._total_size / 1024.0 ** 3)
+        content_length = self._buf.tell()
+        start = self._total_size - content_length
+        stop = self._total_size - 1
+        if stop - start + 1 == content_length:
```

Comment:
What is this fidgeting here?

LHS:

- `stop - start + 1`
- `((self._total_size - 1) - (self._total_size - content_length) + 1` (substituting)
- `self._total_size - 1 - self._total_size + content_length + 1` (expand parentheses)
- `content_length` (simplify, cancel out)

So, the LHS is always equal to `content_length`, which is equal to RHS.

What am I missing here?